cmds/ip/ip.go

```diff
@@ -66,25 +69,86 @@ func dev() netlink.Link {
 	return iface
 }
 
-func showips() {
+func showLinks(withAddresses bool) {
 	ifaces, err := netlink.LinkList()
 	if err != nil {
 		log.Fatalf("Can't enumerate interfaces? %v", err)
 	}
+	var buffer bytes.Buffer
+
 	for _, v := range ifaces {
-		addrs, err := netlink.AddrList(v, netlink.FAMILY_ALL)
-		if err != nil {
-			log.Printf("Can't enumerate addresses")
+		l := v.Attrs()
+		buffer.WriteString(fmt.Sprintf("%d: %s: <%s> mtu %d state %s\n", l.Index, l.Name,
+			strings.Replace(strings.ToUpper(fmt.Sprintf("%s", l.Flags)), "|", ",", -1),
+			l.MTU, strings.ToUpper(l.OperState.String())))
+		buffer.WriteString(fmt.Sprintf("    link/%s %s\n", l.EncapType, l.HardwareAddr))
+		if withAddresses {
+			formatLinkAddresses(&buffer, v)
 		}
-		log.Printf("%v: %v", v, addrs)
 	}
+	fmt.Print(buffer.String())
+}
+
+func formatLinkAddresses(buffer *bytes.Buffer, link netlink.Link) {
+	addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		log.Printf("Can't enumerate addresses")
+	}
+
+	for _, addr := range addrs {
+
+		var scope string
+		switch addr.Scope {
+		case int(netlink.SCOPE_UNIVERSE):
+			scope = "global"
+		case int(netlink.SCOPE_HOST):
+			scope = "host"
+		case int(netlink.SCOPE_SITE):
+			scope = "site"
+		case int(netlink.SCOPE_LINK):
+			scope = "link"
+		}
+
+		var inet string
+		if getIPVersion(addr) == 4 {
```

Comment:
Since this is only used once, I'd prefer this here:

```go
switch len(addr.IPNet.IP) {
case 4:
  inet = "inet"
case 16:
  inet = "inet6"
default:
  panic("welp that sucks.")
}
```